### PR TITLE
feat(error): introduce error types

### LINKED ISSUE #1194: Introduce different error types
For example, to be able to do
```js
error instanceof NotEnoughBalanceError
```
instead of https://github.com/aeternity/superhero-wallet/commit/d7ee86729ec803fea26531d972033855c72c282c#diff-ead5c468661f39fe883b9b2c6f3c4f34d88d09b7ba32b47adbb056bd9ef1d75bR23-R24

### PR DESCRIPTION
**Pull request by subhod-i**
Fix #1194 

Added custom error classes for the account, ae, chain, contract, typings, index, and node files.
Remaining: channel, node-pool, tx, and the utils.

Incorporated all the changes requested from the stale PR #1339

### CODE REVIEW

**Comment by subhod-i** [comment]
@davidyuk codecov is failing. Shall I add unit tests for all the error classes?

File: src/account/multiple.js [hunk 1]
```diff
@@ -134,17 +138,21 @@ export default AccountBase.compose(AsyncInit, {
...
           case 'string':
             decode(account, 'ak')
-            if (!this.accounts[account]) throw new Error(`Account for ${account} not available`)
+            if (!this.accounts[account]) throw new UnavailableAccountError(account)
             return this.accounts[account]
           case 'object':
             return isAccountBase(account) ? account : MemoryAccount({ keypair: account })
           default:
-            throw new Error(`Unknown account type: ${typeof account} (account: ${account})`)
+            throw new TypeError(
```
Comment (mradkov) [inline src/account/multiple.js:153]: Shall this also be exctracted?

**Comment by davidyuk** [comment]
> Shall I add unit tests for all the error classes?

No, they are quite declarative

File: src/ae/aens.js [hunk 2]
```diff
@@ -200,7 +204,7 @@ async function query (name, opt = {}) {
     },
     revoke: async (options = {}) => this.aensRevoke(name, { ...opt, ...options }),
     extendTtl: async (nameTtl = NAME_TTL, options = {}) => {
-      if (!nameTtl || typeof nameTtl !== 'number' || nameTtl > NAME_TTL) throw new Error('Ttl must be an number and less then 180000 blocks')
+      if (!nameTtl || typeof nameTtl !== 'number' || nameTtl > NAME_TTL) { throw new IllegalArgumentError('Ttl must be an number and less then 180000 blocks') }
```
Comment (davidyuk) [inline src/ae/aens.js:207]:
```suggestion
      if (!nameTtl || typeof nameTtl !== 'number' || nameTtl > NAME_TTL) throw new IllegalArgumentError('Ttl must be an number and less then 180000 blocks')
```
avoid extra changes

File: src/utils/error.ts [hunk 3]
```diff
@@ -0,0 +1,428 @@
+/**
```
Comment (davidyuk) [inline src/utils/error.ts:1]: I would rename this file to `errors.ts`

File: src/utils/error.ts [hunk 4]
```diff
@@ -0,0 +1,428 @@
+/**
+ * Error module
+ * @module @aeternity/aepp-sdk/es/utils/error
+ * @example import { BytecodeMismatchError } from '@aeternity/aepp-sdk'
```
Comment (davidyuk) [inline src/utils/error.ts:4]: Don't forget to export these errors in `index.js`

**Review by davidyuk — APPROVED** [review]
This requires some additional rearranging, but it is mergeable as the first step

File: src/utils/error.ts [hunk 5]
```diff
@@ -0,0 +1,428 @@
...
+}
+
+export class UnknownChannelStateError extends AeError {
+  constructor () {
+    super('State Channels FSM entered unknown state')
+    this.name = 'UnknownChannelStateError'
+  }
+}
+
+export class ChannelPingTimedOutError extends AeError {
```
Comment (davidyuk) [inline src/utils/error.ts:380]: We can make hierarchy stronger, for example, inheriting ChannelPingTimedOutError from ChannelError and inheriting ChannelError from AeError, but the hierarchy should be reasonable

**Comment by subhod-i** [comment]
@davidyuk  @mradkov The errors `TypeError, IllegalArgumentError and, MissingParamError` can be potentially eliminated when moving to typescript.

File: src/account/memory.js [hunk 6]
```diff
@@ -44,21 +45,21 @@ export default AccountBase.compose({
   init ({ keypair, gaId }) {
     this.isGa = !!gaId
     if (gaId) {
-      if (!isAddressValid(gaId)) throw new Error('Invalid GA address')
+      if (!isAddressValid(gaId)) throw new InvalidGAaddressError()
```
[the remark this answers is not in the document]
Comment (subhod-i) [reply src/account/memory.js]: can we handle this by fixing merge conflicts during whichever PR gets merged last?

File: src/utils/error.ts (hunk 5 above)
Comment (subhod-i) [reply src/utils/error.ts:555]: Made following as upper hierarchy errors
```
AccountError
AensError
AeppError
ChannelError
CompilerError
ContractError
CryptographyError
NodeError
SwaggerError
TransactionError
WalletError
```

File: src/account/memory.js (hunk 6 above)
[the remark this answers is not in the document]
Comment (davidyuk) [reply src/account/memory.js]: yes, sure

File: src/utils/error.ts [hunk 7]
```diff
@@ -0,0 +1,428 @@
...
+    this.name = 'TypeError'
+  }
+}
+
+export class MissingParamError extends AeError {
+  constructor (msg: string) {
+    super(msg)
+    this.name = 'MissingParamError'
+  }
+}
```
[the remark this answers is not in the document]
Comment (subhod-i) [reply src/utils/error.ts:84]: trying to achieve this using class expressions.

**Comment by davidyuk** [comment]
Why wrapping with specific error classes (that extends Error) is a breaking change?

Proposing to fixup "chore: fix typos in the code", "chore: add missing error classes" to commits where they were introduced.

**Comment by subhod-i** [comment]
> Why wrapping with specific error classes (that extends Error) is a breaking change?
> 
> Proposing to fixup "chore: fix typos in the code", "chore: add missing error classes" to commits where they were introduced.

There are many changes in my local now too based on review suggestions. there will be few of such chore commits.
Can we squash and merge to develop as a single commit `feat(error)!: introduce error class`?

File: src/ae/aens.js (hunk 2 above)
Comment (subhod-i) [reply src/ae/aens.js:206]: @davidyuk  Linter throws an error here. moving the statement to next line without braces

**Review by davidyuk — APPROVED** [review]
I'm fine to squash merge it, but without marking as a breaking change

**Comment by subhod-i** [comment]
> I'm fine to squash merge it, but without marking as a breaking change

So it's not a breaking change. In fact, the existing try-catch blocks will work as it is.

**Comment by subhod-i** [comment]
@davidyuk ready to merge

**Review by mradkov — APPROVED** [review]

**Review by thepiwo — APPROVED** [review]
cool!

**Pull request merged**